File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/api/ValorantSkinController.java
package fr.huiitre.tools.modules.riot.valorant.api;

import java.util.List;
import java.util.UUID;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import fr.huiitre.tools.modules.core.common.api.RequiredRole;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.usecase.GetValorantSkinByAssetIdUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.usecase.GetValorantSkinByLevelUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.usecase.GetValorantSkinUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.usecase.ListValorantSkinsUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.usecase.ListValorantSkinsByThemeUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.skin.view.ValorantSkinView;

@RestController
@RequestMapping("/riot/valorant")
public class ValorantSkinController {

    private final ListValorantSkinsUseCase listValorantSkinsUseCase;
    private final GetValorantSkinUseCase getSkinUseCase;
    private final GetValorantSkinByAssetIdUseCase getSkinByAssetIdUseCase;
    private final GetValorantSkinByLevelUseCase getSkinByLevelUseCase;
    private final ListValorantSkinsByThemeUseCase listSkinsByThemeUseCase;

    public ValorantSkinController(
            ListValorantSkinsUseCase listValorantSkinsUseCase,
            GetValorantSkinUseCase getSkinUseCase,
            GetValorantSkinByAssetIdUseCase getSkinByAssetIdUseCase,
            GetValorantSkinByLevelUseCase getSkinByLevelUseCase,
            ListValorantSkinsByThemeUseCase listSkinsByThemeUseCase) {
        this.listValorantSkinsUseCase = listValorantSkinsUseCase;
        this.getSkinUseCase = getSkinUseCase;
        this.getSkinByAssetIdUseCase = getSkinByAssetIdUseCase;
        this.getSkinByLevelUseCase = getSkinByLevelUseCase;
        this.listSkinsByThemeUseCase = listSkinsByThemeUseCase;
    }

    @RequiredRole(RoleCode.READ_ONLY)
    @GetMapping("/skins")
    public List<ValorantSkinView> listSkins() {
        return listValorantSkinsUseCase.execute();
    }

    @RequiredRole(RoleCode.READ_ONLY)
    @GetMapping("/skins/{id}")
    public ValorantSkinView getSkin(@PathVariable Long id) {
        return getSkinUseCase.execute(id);
    }

    @RequiredRole(RoleCode.READ_ONLY)
    @GetMapping("/skins/by-asset/{assetId}")
    public ValorantSkinView getSkinByAssetId(@PathVariable UUID assetId) {
        return getSkinByAssetIdUseCase.execute(assetId);
    }

    @RequiredRole(RoleCode.READ_ONLY)
    @GetMapping("/skins/by-level/{levelAssetId}")
    public ValorantSkinView getSkinByLevel(@PathVariable UUID levelAssetId) {
        return getSkinByLevelUseCase.execute(levelAssetId);
    }

    @RequiredRole(RoleCode.READ_ONLY)
    @GetMapping("/skins/by-theme/{themeUuid}")
    public List<ValorantSkinView> listSkinsByTheme(@PathVariable UUID themeUuid) {
        return listSkinsByThemeUseCase.execute(themeUuid);
    }
}

File: api/Modules/Riot/Valorant/Api/ValorantSkinController.cs
using Microsoft.AspNetCore.Mvc;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Usecases;
using Tools.Api.Modules.Riot.Valorant.Application.Skin.Views;

namespace Tools.Api.Modules.Riot.Valorant.Api;

// Catalogue des skins. accountId est facultatif partout : il ne sert qu'à renseigner « possédé »
// et « suivi » sur chaque skin.
[ApiController]
[Route("riot/valorant")]
public class ValorantSkinController : ControllerBase
{
    [HttpGet("skins")]
    public Task<List<ValorantSkinView>> ListSkins(
        [FromServices] ListValorantSkinsUseCase listValorantSkinsUseCase,
        [FromQuery] long? accountId
    )
    {
        return listValorantSkinsUseCase.Execute(accountId);
    }

    [HttpGet("skins/{id:long}")]
    public Task<ValorantSkinView> GetSkin(
        [FromServices] GetValorantSkinUseCase getValorantSkinUseCase,
        [FromRoute] long id,
        [FromQuery] long? accountId
    )
    {
        return getValorantSkinUseCase.Execute(id, accountId);
    }

    [HttpGet("skins/by-asset/{assetId:guid}")]
    public Task<ValorantSkinView> GetSkinByAssetId(
        [FromServices] GetValorantSkinByAssetIdUseCase getValorantSkinByAssetIdUseCase,
        [FromRoute] Guid assetId,
        [FromQuery] long? accountId
    )
    {
        return getValorantSkinByAssetIdUseCase.Execute(assetId, accountId);
    }

    // Le storefront Riot ne renvoie que des UUID de *levels*, jamais l'UUID du skin racine :
    // c'est cette route qui fait le pont.
    [HttpGet("skins/by-level/{levelAssetId:guid}")]
    public Task<ValorantSkinView> GetSkinByLevel(
        [FromServices] GetValorantSkinByLevelUseCase getValorantSkinByLevelUseCase,
        [FromRoute] Guid levelAssetId,
        [FromQuery] long? accountId
    )
    {
        return getValorantSkinByLevelUseCase.Execute(levelAssetId, accountId);
    }

    [HttpGet("skins/by-theme/{themeUuid:guid}")]
    public Task<List<ValorantSkinView>> ListSkinsByTheme(
        [FromServices] ListValorantSkinsByThemeUseCase listValorantSkinsByThemeUseCase,
        [FromRoute] Guid themeUuid,
        [FromQuery] long? accountId
    )
    {
        return listValorantSkinsByThemeUseCase.Execute(themeUuid, accountId);
    }
}
